### ISSUE TITLE
OneSignal Flutter SDK Version 3 - Beta! 

### ISSUE THREAD

**Issue by Jeasmine** [OneSignal/OneSignal-Flutter-SDK#357]
Hello everyone!
The Flutter major release version 3 is now in **Beta**!

This version includes many exciting changes including an overall simplification of the API as well as many of the underlying native improvements and bug fixes.

Some key changes already in Beta:

* Better notification control while app is in foreground
* Part of OneSignal initialization is now done for you automatically. Simply set the app id to complete the process.

## Why you should start migrating now
Most importantly, we want to give you the opportunity to voice your opinion and provide feedback so that we can make changes as needed to better suit your implementation.

Furthermore, it is always beneficial to get a head start on migrating your implementation to ensure a smooth transition. As we continue improving OneSignal, being fully migrated to the latest version of the SDK will help minimize potential issues and ensure you can focus on providing value to your customers, improving your business, and creating a delightful messaging experience.

To begin migrating your app to version 3, read our migration guide [**here**](https://documentation.onesignal.com/docs/step-by-step-flutter-2x-to-300-upgrade-guide). Regarding this guide, please let us know if anything could be clarified or communicated better.

[Releases Page](https://github.com/OneSignal/OneSignal-Flutter-SDK/releases)

We are looking forward to hearing from you.

Enjoy!

**Comment by pr-1** [comment]
Is Null Safety included ?

**Comment by laurent-humbert** [comment]
> Is Null Safety included ?

dart pub outdated --mode=null-safety

says no

 :-(

**Comment by johnsheehan** [comment]
After following the migration guide, I am getting the following 

```
Failed to build iOS app
Error output from Xcode build:
↳
    ** BUILD FAILED **
Xcode's output:
↳
    OneSignalNotificationServiceExtension/NotificationService.m:1:9: fatal error: 'OneSignal/OneSignal.h' file not found
    #import <OneSignal/OneSignal.h>
            ^~~~~~~~~~~~~~~~~~~~~~~
    1 error generated.
    note: Using new build system
    note: Building targets in parallel
    note: Planning build
    note: Constructing build description
Could not build the application for the simulator.
```

Is there an updated version of `NotificationService.m` to use?

**Comment by superakabo** [comment]
@Jeasmine 
OneSignal.shared.setNotificationOpenedHandler((OSNotificationOpenedResult result) {}
is not triggered. It works in version 2.6.3 though.

**Comment by MeAtPros** [comment]
When I click on the push notification, setNotificationOpenedHandler should be called
```
OneSignal.shared.setNotificationOpenedHandler((OSNotificationOpenedResult result) {
    try {
      ReceiveNotification.instance.addPushData(result.notification);
    }
    catch(e) {
      print('e : ' + e.toString());
    }
  });
```

Unfortunately, this does not happen and I get the error below

```
I/WM-WorkerWrapper( 7882): Worker result FAILURE for Work [ id=eacb68a2-f7f4-4dd6-9be0-785162b87877, tags={ com.onesignal.OSNotificationRestoreWorkManager$NotificationRestoreWorker } ]
W/OneSignal( 7882): FusedLocationApi.requestLocationUpdates failed!
W/OneSignal( 7882): java.lang.NullPointerException: Calling thread must be a prepared Looper thread.
W/OneSignal( 7882): 	at com.google.android.gms.common.internal.Preconditions.checkNotNull(Unknown Source)
W/OneSignal( 7882): 	at com.google.android.gms.internal.location.zzq.requestLocationUpdates(Unknown Source)
W/OneSignal( 7882): 	at com.onesignal.GMSLocationController$FusedLocationApiWrapper.requestLocationUpdates(GMSLocationController.java:198)
W/OneSignal( 7882): 	at com.onesignal.GMSLocationController$LocationUpdateListener.init(GMSLocationController.java:181)
W/OneSignal( 7882): 	at com.onesignal.GMSLocationController$LocationUpdateListener.<init>(GMSLocationController.java:165)
W/OneSignal( 7882): 	at com.onesignal.GMSLocationController.onFocusChange(GMSLocationController.java:117)
W/OneSignal( 7882): 	at com.onesignal.LocationController.onFocusChange(LocationController.java:326)
W/OneSignal( 7882): 	at com.onesignal.OneSignal.onAppLostFocus(OneSignal.java:1203)
W/OneSignal( 7882): 	at com.onesignal.ActivityLifecycleHandler$AppFocusRunnable.run(ActivityLifecycleHandler.java:282)
W/OneSignal( 7882): 	at java.lang.Thread.run(Thread.java:762)
```

**Comment by MeAtPros** [comment]
Same issues was fixed on OneSignal Android SDK - notifications omitted in OSRemoteNotificationReceivedHandler and from disabled channels being restored and shown as silent notifications. PR #1289, Issue #1265

**Comment by MeAtPros** [comment]
@Jeasmine Only Android

**Comment by superakabo** [comment]
> > @Jeasmine
> > OneSignal.shared.setNotificationOpenedHandler((OSNotificationOpenedResult result) {}
> > is not triggered. It works in version 2.6.3 though.
> 
> Thanks for the feedback. Is it not working for both Android and iOS?

Yes, it does not work for both Android and iOS.

**Comment by MeAtPros** [comment]
@Jeasmine Any update with beta-1 ? I need a fix for this problem fairly quickly as it prevents me from the released versions. Unfortunately, I can't revert my project to the previous version.

**Comment by Jeasmine** [comment]
@MeAtPros, we are working on the new release, hopefully, will be out this week

**Comment by Jeasmine** [comment]
@MeAtPros there is this PR up https://github.com/OneSignal/OneSignal-Flutter-SDK/pull/375, but unfortunately, the scenario when the app is swiped away the open handler is not being called, there is a problem with flutter integration still trying to figure it out.

**Comment by edsonboldrini** [comment]
I'm looking for this too, the beta release solves a critical problem in iOS that in version 2 I can't supress default launch url behavior to only handle inside my app, and the version 3 helped me to do this :)

**Comment by marclester-tang** [comment]
Same having issues with launch url, we're unable to deep link into our app with the launch_url parameter

**Comment by phcs971** [comment]
After migrating we are having the problem in #374 which is preventing us from distributing the app.

**Comment by nickscamara** [comment]
I am having the same problem as #374. I tried reverting OneSignal to 2.6.3 but I still get the same errors...

**Comment by MeAtPros** [comment]
@Jeasmine @mikechoch @mpillar Can you set a date when a proper release of the update can be expected?

**Comment by Jeasmine** [comment]
Beta 2 has been release! https://github.com/OneSignal/OneSignal-Flutter-SDK/releases/tag/3.0.0-beta2

**Comment by gilthonweapps** [comment]
Hi, the issue with the external user id being randomly set when using setExternalUserId is still there.

**Comment by Jeasmine** [comment]
@gilthonweapps, Could you open a GitHub issue with the steps to reproduce and the description, or could you point me out to the issue if there is already one created? Thanks!

**Comment by gilthonweapps** [comment]
Hi @Jeasmine  it is issue #150

**Comment by Jeasmine** [comment]
> Hi @Jeasmine, it is issue #150

Thanks for pointing this up. This is something happening on the native side. We will handle this on iOS SDK.

**Comment by Jeasmine** [comment]
Hey all! Beta 3 has been released https://github.com/OneSignal/OneSignal-Flutter-SDK/releases/tag/3.0.0-beta3. We included some native side bug fixes and additional-data parsing bug under android. Thanks again for the feedback!

**Comment by binish784** [comment]
support for setSubscription method in v3 ?

**Comment by binish784** [comment]
support for setSubscription method in v3 ?

**Comment by youssefali424** [comment]
iOs cold start problem is fixed?

**Comment by enricobenedos** [comment]
Good morning,

I'm trying to migrate my app and I'm using OneSignal Flutter SDK from years. It seems that on documentation it is missing how to substitute `setSubscription(true)`. Can someone please explain me how to substitute it?

So it seems that to substitute `setSubscription(false)` it is necessary to use `disablePush()`.

So now when I call `init(myAppId)` this error message is returned:
```log
W/OneSignal( 6609): appContext set, but please call setAppId(appId) with a valid appId to complete OneSignal init!
```

**Comment by leonardodincalci** [comment]
Hello, with Intel Mac and Flutter 2.2.0 i got this error:

```
ld: in /Users/_/git/_/ios/Pods/OneSignal/iOS_SDK/OneSignalSDK/Framework/OneSignal.framework/OneSignal(OneSignal-arm64-master.o), building for iOS Simulator, but linking in object file built for iOS, file '/Users/_/git/_/ios/Pods/OneSignal/iOS_SDK/OneSignalSDK/Framework/OneSignal.framework/OneSignal'
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by macortesn** [comment]
Hi, iI am having this issue when running the application in IOS. Does anyone know how I can solve this?

![image](https://user-images.githubusercontent.com/21014156/120410398-843aa780-c318-11eb-81c4-4ce37a635cdc.png)

**Comment by Jeasmine** [comment]
> Good morning,
> 
> I'm trying to migrate my app and I'm using OneSignal Flutter SDK from years. It seems that on documentation it is missing how to substitute `setSubscription(true)`. Can someone please explain me how to substitute it?
> 
> So it seems that to substitute `setSubscription(false)` it is necessary to use `disablePush()`.
> 
> So now when I call `init(myAppId)` this error message is returned:
> 
> ```
> W/OneSignal( 6609): appContext set, but please call setAppId(appId) with a valid appId to complete OneSignal init!
> ```

Hey @enricobenedos sry for the late response, have you been able to check the migration guide https://documentation.onesignal.com/docs/step-by-step-flutter-2x-to-300-upgrade-guide? Were you able to migrate?

**Comment by Jeasmine** [comment]
Hey @leonardodincalci and @macortesn, which beta version are you guys using? The last beta cut version 4 has changed iOS dependency to XCFramework, you will need to call _pod install_ again, if that don't work call _pod deintegrate_ and _pod install_. Let me know if that helps!

**Comment by Jeasmine** [comment]
The Flutter SDK is now out of Beta!
---

Please upgrade to version `^3.0.0`
Enjoy!

**Comment by Jeasmine** [comment]
@macortesn did you tried pod repo update or pod install --repo-update? cc: @emawby

**Comment by macortesn** [comment]
@Jeasmine I updated to the latest version, cleaned the project, ran the command pod install --repo-update, tried other things and the error persists

**Comment by jkasten2** [comment]
@macortesn The error in your screenshot:
```
+[OneSignal initWithLaunchOptions:]: unrecognized selector sent to class...
```
means that the `OneSignal` iOS library is on a 2.x.x version where it should be on a 3.x.x version. Can you check both your `Podfile` and your `Podfile.lock` for `OneSignal` and make sure it is 3.0.0 or newer?

**Comment by macortesn** [comment]
@jkasten2 I verified  and the version of the library is the latest. I made the following change:
```
target 'OneSignalNotificationServiceExtension' do
  use_frameworks!
  pod 'OneSignal', '>= 2.9.3', '< 3.0'
end
 ```
 To
```
target 'OneSignalNotificationServiceExtension' do
  use_frameworks!
  pod 'OneSignal', '>= 3.0'
end
```
With this change I got the error:
```
Thread 1: "+[OneSignal addSMSSubscriptionObserver:]: unrecognized selector sent to class 0x10aa790a8"
```
In the OneSignalPlugin.m file is originated the error. In this [line](https://github.com/OneSignal/OneSignal-Flutter-SDK/blob/f9da7026aae73ad99051058b49504c70416185d6/ios/Classes/OneSignalPlugin.m#L102) show the warning 
```
Incompatible pointer types sending 'OneSignalPlugin *' to parameter of type 'NSObject<OSSMSSubscriptionObserver> * _Nonnull'
```
So I comment the line of the error and the application ran correctly.

**Comment by mingosnunes** [comment]
I have this error on the open handler trigger -> [issue](https://github.com/OneSignal/OneSignal-Flutter-SDK/issues/407)

Thanks

**Comment by MeAtPros** [comment]
Still this problem has not been fixed and we have been waiting for a couple of weeks https://github.com/OneSignal/OneSignal-Flutter-SDK/issues/372

**Comment by enricobenedos** [comment]
> > Good morning,
> > I'm trying to migrate my app and I'm using OneSignal Flutter SDK from years. It seems that on documentation it is missing how to substitute `setSubscription(true)`. Can someone please explain me how to substitute it?
> > So it seems that to substitute `setSubscription(false)` it is necessary to use `disablePush()`.
> > So now when I call `init(myAppId)` this error message is returned:
> > ```
> > W/OneSignal( 6609): appContext set, but please call setAppId(appId) with a valid appId to complete OneSignal init!
> > ```
> 
> Hey @enricobenedos sry for the late response, have you been able to check the migration guide https://documentation.onesignal.com/docs/step-by-step-flutter-2x-to-300-upgrade-guide? Were you able to migrate?

I've already check the migration guide. First it was possibile to initialise the SDK and then to subscribe to notification. How it is possible now? What is the replacement for the `setSubscription()` method?

In the migration guide the `setSubscription()` replacement is not explained.

**Comment by Jeasmine** [comment]
> https://documentation.onesignal.com/docs/step-by-step-flutter-2x-to-300-upgrade-guide

Yeah you are correct `disablePush()` is the replacement for `setSubscription()`, this will be updated in the documentation. Thanks for pointing it up.

**Comment by malvinpratama** [comment]
i have issue with setNotificationWillShowInForegroundHandler.
setNotificationWillShowInForegroundHandler cannot handle the data if the app is out of focus.
if use sdk < 3.0.0 it's work with setNotificationReceivedHandler.
in setNotificationReceivedHandler you can handle the data even if the app is out of focus.
but in setNotificationWillShowInForegroundHandler your app must be in focus.
i don't know it's an SDK issue or the documentation to handle cases like this doesn't exist but can handle cases like this.

**Comment by rafampessoa** [comment]
the same issue that @leonardodincalci reported. Have you found an workaround i'm unable to run my app ):

**Comment by jonchui** [comment]
@Jeasmine I'm using 3.1.0 in a flutter app and iOS works fine. But when I build for android, i get the weirdest error, when running `flutter run`  (which on android, is running `./gradlew --info :app:assembleDebug`):

```
> In project 'app' a resolved Google Play services library dependency depends on another at an exact version (e.g. "[19.0.
  0, 22.0.99]", but isn't being resolved to that version. Behavior exhibited by the library will be unknown.

  Dependency failing: com.onesignal:OneSignal:4.4.1 -> com.google.firebase:firebase-messaging@[19.0.0, 22.0.99], but fireb
  ase-messaging version was 22.0.0.
```

I thought the `[` meant a range of versions? Of which i'm pretty sure `22.0.0` is between 19.0.0 and `22.0.99` ?!?! 

Anyways - I'll file a ticket for this too - but just wondering if anyone else was getting this... 
seems that 

From the text error, it seems that android build system is mistakenly thinking that `@[19.0.0, 22.0.99]` is an `exact version` ?

**Comment by jkasten2** [comment]
Since the OneSignal-Flutter-SDK 3.0.0 has been fully released closing and locking this issue. Please create a new issue instead.
